### PR TITLE
Progress cancellations

### LINKED ISSUE #1494: Allow cancellation of uploads, downloads and transfer.
Currently there is no way of cancelling these operations while in progress.

Use the toasts to show a `cancellation` button,
show loading while the operation gets cancelled!

### PR DESCRIPTION
**Pull request by tanmoyAtb**
closes #1494

### CODE REVIEW

**Review by RyRy79261 — APPROVED** [review]

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 1]
```diff
@@ -507,73 +525,97 @@ const FilesProvider = ({ children }: FilesContextProps) => {
...
+          link.href = URL.createObjectURL(new Blob([zipped]))
+          link.download = "archive.zip"
+          link.click()
+          setDownloadsInProgress(false)
 
+          updateToast(toastId, {
+            title: t`Download Complete`,
+            type: "success",
+            progress: undefined,
+            onProgressCancel: undefined
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx]: Can we add the `isClosable: true` here?
```suggestion
            onProgressCancel: undefined,
            isClosable: true
```

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 2]
```diff
@@ -507,73 +525,97 @@ const FilesProvider = ({ children }: FilesContextProps) => {
...
+          let errorMessage = t`Downloads failed`
+          if (axios.isCancel(error)) {
+            errorMessage = t`Downloads cancelled`
+          }
+          setDownloadsInProgress(false)
+          updateToast(toastId, {
+            title: errorMessage,
+            type: "error",
+            progress: undefined,
+            onProgressCancel: undefined
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx]: And here
```suggestion
            onProgressCancel: undefined,
            isClosable: true
```

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 3]
```diff
@@ -591,17 +633,23 @@ const FilesProvider = ({ children }: FilesContextProps) => {
       updateToast(toastId, {
         title: t`Download Complete`,
         type: "success",
-        progress: undefined
+        progress: undefined,
+        onProgressCancel: undefined
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx]: and here
```suggestion
        onProgressCancel: undefined,
        isClosable: true
```

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 4]
```diff
@@ -591,17 +633,23 @@ const FilesProvider = ({ children }: FilesContextProps) => {
...
+      if (axios.isCancel(error)) {
+        errorMessage = t`Downloads cancelled`
+      }
       updateToast(toastId, {
-        title: `${t`An error occurred: `} ${typeof(error) === "string" ? error : error.length ? error[0].message : ""}`,
+        title: errorMessage,
         type: "error",
-        progress: undefined
+        progress: undefined,
+        onProgressCancel: undefined
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx]:
```suggestion
        onProgressCancel: undefined,
        isClosable: true
```

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 5]
```diff
@@ -686,10 +734,16 @@ const FilesProvider = ({ children }: FilesContextProps) => {
...
+    const cancelToken = cancelSource.token
+
     const toastParams: ToastParams = {
       title: t`Sharing your file (Downloading)`,
       type: "success",
-      progress: 0
+      progress: 0,
+      toastPosition: "bottomRight",
+      onProgressCancel: cancelSource.cancel,
+      isClosable: false
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx:749]: I think any toast should be closable, beside the ones from upload/download/sharing
```suggestion
      isClosable: true
```

File: packages/files-ui/src/Contexts/FilesContext.tsx [hunk 6]
```diff
@@ -743,11 +800,16 @@ const FilesProvider = ({ children }: FilesContextProps) => {
...
+      if (axios.isCancel(error)) {
+        errorMessage = t`Sharing cancelled`
+      }
       updateToast(toastId, {
-        title: `${t`An error occurred: `} ${typeof(error) === "string" ? error : error.length ? error[0].message : ""}`,
+        title: errorMessage,
         type: "error",
-        progress: undefined
+        progress: undefined,
+        onProgressCancel: undefined
```
Comment (Tbaut) [inline packages/files-ui/src/Contexts/FilesContext.tsx]:
```suggestion
        onProgressCancel: undefined,
        isClosable: true
```

**Review by Tbaut — APPROVED** [review]
This is really good, you're on :fire: !
Made some suggestions to have the toasts closable as much as possible.

File: packages/files-ui/src/Contexts/FilesContext.tsx (hunk 1 above)
Comment (tanmoyAtb) [reply packages/files-ui/src/Contexts/FilesContext.tsx]: Yes makes sense, after operations are complete!

File: packages/files-ui/src/Contexts/FilesContext.tsx (hunk 5 above)
Comment (tanmoyAtb) [reply packages/files-ui/src/Contexts/FilesContext.tsx:749]: Yes, this one falls under sharing, this is the start of downloading part!

**Comment by tanmoyAtb** [comment]
> This is really good, you're on 🔥 !
> Made some suggestions to have the toasts closable as much as possible.

Hey there! thanks for the suggestions, I updated thee toasts to be closable whenever operations complete!

**Review by Tbaut — APPROVED** [review]
:fire: :fire: :fire: :fire:

**Pull request merged**